zerver/lib/actions.py

```diff
@@ -1888,20 +1888,27 @@ def do_send_typing_notification(
         realm: Realm,
```

Comment:
I would prefer to just have a separate action function for stream notifications, so that you don't need the optional parameters and the conditional logic.  Two simple functions are better than one complex one, and we're not really sharing any useful logic here.

I would also consider just having a new `type` (no pun intended) for the event with the value `stream_typing`, so that the documentation piece is easier here (again, no Union types would be needed).

Reply:
Yes, having a different type would make it simpler and separating the function will make it seem less complex. Will make the required changes.

(Thought I'd make the changes and ask for another review soon but not I'm getting a chance to work on this as most of my time is going into academics. Leaving an update since it's been 2weeks..:)

Reply:
Yeah, I think the separate function makes sense, with a bit of preparatory refactoring to avoid duplicating code.

Reply:
Separated the function. There wasn't much duplicate code. 

Also added `type` `stream_typing`.